metaspace/browser/sm/browser/mz_search.py

```
@@ -0,0 +1,99 @@
+
+
+class S3File(BinaryFile):
+    def __init__(self, s3_object, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.s3_object = s3_object
+        self.position: int = 0
+
+    @property
+    def size(self) -> int:
+        return self.s3_object.content_length
+
+    def seek(self, offset: int, **kwargs) -> int:
+        assert 0 <= offset < self.size, f"0 <= {offset} < {self.size}"
+
+        self.position = offset
+        return self.position
+
+    def read(self, n: int) -> bytes:
+        assert n > 0
+
+        start, end = self.position, self.position + n - 1  # end included
+        if end > self.size - 1:
+            end = self.size - 1
+        range_header = f"bytes={start}-{end}"
+        return self.s3_object.get(Range=range_header)["Body"].read()
+
+
+def search_and_fetch_mz_peaks(
+    stream: BinaryFile, mz_index: np.ndarray, mz_lo: float, mz_hi: float
+) -> np.ndarray:
+    mz_lo_chunk_idx, mz_hi_chunk_idx = np.searchsorted(mz_index, [mz_lo, mz_hi])
+    mz_lo_chunk_idx -= 1  # previous chunk actually includes value
+
+    offset = mz_lo_chunk_idx * CHUNK_SIZE
+    stream.seek(offset)
+    bytes_to_read = (mz_hi_chunk_idx - mz_lo_chunk_idx + 1) * CHUNK_SIZE
+    bytes = stream.read(bytes_to_read)
+    mz_chunks_array = np.frombuffer(bytes, dtype="f").reshape(-1, 3)
+    idx_lo, idx_hi = np.searchsorted(mz_chunks_array[:, 0], [mz_lo, mz_hi])
+    mz_peaks = mz_chunks_array[idx_lo:idx_hi]  # idx_hi equals to index after last
+
+    return mz_peaks
+
+
+def create_mz_image(mz_peaks: np.ndarray, coordinates: np.ndarray) -> np.ndarray:
+    min_x, min_y = np.amin(coordinates, axis=0)
+    max_x, max_y = np.amax(coordinates, axis=0)
+    nrows, ncols = max_y - min_y + 1, max_x - min_x + 1
+    mz_image = np.zeros(shape=(nrows, ncols))
+
+    image_coords = coordinates[mz_peaks[:, 2].astype(int)]
+    xs = image_coords[:, 0] - min_x
+    ys = image_coords[:, 1] - min_y
+    mz_image[ys, xs] += mz_peaks[:, 1]
```

Unfortunately this doesn't behave as you'd expect. If multiple values exist for the same pixel, this type of assignment will cause the last one to be selected and the rest to be discarded. e.g.
```python
image = np.zeros((3,3))
xs, ys = np.array([0,0,0,0]), np.array([0,1,0,1])
ints = np.array([1,2,3,5])
image[xs, ys] += ints
image

> array([[3., 5., 0.],    # Would expect [4., 7,. 0.],
>        [0., 0., 0.],
>        [0., 0., 0.]])
```

I'd suggest using `coo_matrix.todense()` if possible. If it's too slow, a Numba njit function would let you efficiently loop through the whole array doing `+=` with each item.

---

Thanks, that's a good catch. I didn't test this extensively so it was pretty easy to overlook issues that don't have an immediate effect. Weird that numpy silently replaces `+=` operator with `=`.

---

It doesn't change it to `=`, it just does something weird so that a maximum of one source element is `+=`'d to each destination index. I suspect that it does something like this:
```
# Original code:
a[xs, ys] += b
# Actual operations
temp1 = np.zeros_like(a)
temp1[xs, ys] = b
a += temp1
```

---

Yeah, I understand. I meant that effectively `image[ys, xs] += ints` produces the same result as `image[ys, xs] = ints`
```
In[4]: image = np.zeros((3,3))
  ...: xs, ys = np.array([0,0,0,0]), np.array([0,1,0,1])
  ...: ints = np.array([1,2,3,5])
  ...: image[xs, ys] = ints
  ...: image
Out[4]: 
array([[3., 5., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])
```

---

```
np.add.at(mz_image, [ys, xs], mz_peaks[:, 1])
```
Seems to fix the mentioned issue.